source/particle/world.cc

```diff
@@ -1211,32 +1211,71 @@ namespace aspect
 
       const FiniteElement<dim> &velocity_fe = this->get_fe().base_element(this->introspection().base_elements.velocities);
 
+      const bool compute_fluid_velocity = this->include_melt_transport() && property_manager->get_data_info().fieldname_exists("melt_presence");
+      unsigned int fluid_component_index = numbers::invalid_unsigned_int;
+      std::vector<bool> use_fluid_velocity;
+      if (compute_fluid_velocity)
+        {
+          fluid_component_index = this->introspection().variable("fluid velocity").first_component_index;
+          const unsigned int melt_property_index = property_manager->get_data_info().get_position_by_field_name("melt_presence");
+
+          use_fluid_velocity.resize(particles_in_cell,false);
+
+          typename std::multimap<types::LevelInd, Particle<dim> >::iterator it = begin_particle;
+          for (unsigned int particle_index = 0; it!=end_particle; ++it,++particle_index)
+            if (it->second.get_properties()[melt_property_index] == 1.0)
+              use_fluid_velocity[particle_index] = true;
+        }
+
       for (unsigned int j=0; j<velocity_fe.dofs_per_cell; ++j)
         {
-          Tensor<1,dim> solution_at_support_point;
-          Tensor<1,dim> old_solution_at_support_point;
+          Tensor<1,dim> velocity_at_support_point;
+          Tensor<1,dim> old_velocity_at_support_point;
+
+          Tensor<1,dim> fluid_velocity_at_support_point;
+          Tensor<1,dim> old_fluid_velocity_at_support_point;
+
           for (unsigned int dir=0; dir<dim; ++dir)
             {
               const unsigned int support_point_index
-                = this->get_fe().component_to_system_index(/*velocity component=*/ this->introspection().component_indices.velocities[dir],
-                                                                                   /*dof index within component=*/ j);
-              solution_at_support_point[dir] = this->get_solution()[cell_dof_indices[support_point_index]];
-              old_solution_at_support_point[dir] = this->get_old_solution()[cell_dof_indices[support_point_index]];
+                = this->get_fe().component_to_system_index(this->introspection().component_indices.velocities[dir],j);
+
+              velocity_at_support_point[dir] = this->get_solution()[cell_dof_indices[support_point_index]];
+              old_velocity_at_support_point[dir] = this->get_old_solution()[cell_dof_indices[support_point_index]];
+
+              if (compute_fluid_velocity)
```

Comment:
This might be easier to read if you ended the previous `for` loop here and then started from scratch:
```
  if (compute_...)
   for (unsigned int dir=...)
```
(I can't really explain why I find that easier to read other than the fact that I like pulling overarching conditionals out of innermost loops. That is often also faster because it avoids the repeated evaluation of conditions and the introduction of jumps into the innermost loop, but this is surely of no concern at all here.)